lnrpc/routerrpc/router_backend.go

```diff
@@ -358,99 +363,97 @@ func (r *RouterBackend) MarshallRoute(route *route.Route) (*lnrpc.Route, error)
 			PubKey: hex.EncodeToString(
 				hop.PubKeyBytes[:],
 			),
-			TlvPayload: !hop.LegacyPayload,
-			MppRecord:  mpp,
+			CustomRecords: tlvMap,
+			TlvPayload:    !hop.LegacyPayload,
+			MppRecord:     mpp,
 		}
 		incomingAmt = hop.AmtToForward
 	}
 
 	return resp, nil
 }
 
-// UnmarshallHopByChannelLookup unmarshalls an rpc hop for which the pub key is
-// not known. This function will query the channel graph with channel id to
-// retrieve both endpoints and determine the hop pubkey using the previous hop
-// pubkey. If the channel is unknown, an error is returned.
-func (r *RouterBackend) UnmarshallHopByChannelLookup(hop *lnrpc.Hop,
-	prevPubKeyBytes [33]byte) (*route.Hop, error) {
-
-	// Discard edge policies, because they may be nil.
-	node1, node2, err := r.FetchChannelEndpoints(hop.ChanId)
-	if err != nil {
-		return nil, err
-	}
+// UnmarshallCustomRecords unmarshall rpc custom records to tlv records.
+func UnmarshallCustomRecords(rpcRecords map[uint64][]byte) ([]tlv.Record,
+	error) {
 
-	var pubKeyBytes [33]byte
-	switch {
-	case prevPubKeyBytes == node1:
-		pubKeyBytes = node2
-	case prevPubKeyBytes == node2:
-		pubKeyBytes = node1
-	default:
-		return nil, fmt.Errorf("channel edge does not match expected node")
+	if rpcRecords == nil {
+		return nil, nil
 	}
 
-	var tlvRecords []tlv.Record
+	tlvRecords := tlv.MapToRecords(rpcRecords)
 
-	mpp, err := UnmarshalMPP(hop.MppRecord)
-	if err != nil {
-		return nil, err
+	// tlvRecords is sorted, so we only need to check that the first
+	// element is within the custom range.
+	if uint64(tlvRecords[0].Type()) < hop.CustomTypeStart {
```

Comment:
do we know length > 0?

Reply:
Maybe a zero element map from grpc is always nil, but not fully sure. Replaced the nil check at the top by a len()==0 check